kv: add ability to verify pipelined replicated shared/exclusive locks

arulajmani opened the pull request:
Previously, QueryIntent requests were only used to verify whether an
intent was successfully evaluated and replicated. This patch extends
QueryIntent request to also be able to verify whether a pipelined
shared or exclusive lock was successfully replicated or not.

Informs https://github.com/cockroachdb/cockroach/issues/117978

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119933)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 7 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvpb/api.proto` line 1440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj__Rb1tXTjGJtWJ-B:-Nsj__Rb1tXTjGJtWJ-C:bdzybcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvpb/api.proto#L1440)):*
> ```proto
>   // with the supplied lock strength or something stronger at the sequence
>   // number.
>   kv.kvserver.concurrency.lock.Strength lock_strength = 4;
> ```

Do we need to set this field in `resolveIndeterminateCommitForTxnProbe`?
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj_wC7DNsHKwnHnuiX:-Nsj_wC7DNsHKwnHnuiY:b-by4k6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L149)):*
> ```Go
> 			// in a meaningful way; see if we can get rid of it, or, at the least, not
> 			// return locks from other transactions for Shared and Exclusive locks.
> 			return result.Result{}, kvpb.NewIntentMissingError(args.Key, nil /* intent */)
> ```

Should we fix up the `NewIntentMissingError` interface so that we can pass in the missing lock?
___
*[`pkg/roachpb/data.go` line 1004 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjWJ2Y9RMGbLPL0-JQ:-NsjWJ2Y9RMGbLPL0-JR:be3gto1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/roachpb/data.go#L1004)):*
> ```Go
> }
> 
> func (t *Transaction) GetTxnMeta() *enginepb.TxnMeta {
> ```

This doesn't feel like it belongs as a method on `Transaction`. Rather, it feels like it should be a method on `MVCCWriteOptions`.
___
*[`pkg/roachpb/data.proto` line 630 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjVV6HBJSJOSXRmtxi:-NsjVV6HBJSJOSXRmtxj:be0at47) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/roachpb/data.proto#L630)):*
> ```proto
>   // The key that the write was made at.
>   bytes key = 1 [(gogoproto.casttype) = "Key"];
>   // The sequence number of the request that created the write.
> ```

This will only be set if `lock_strength` is Intent, right?

EDIT: or maybe not. Maybe this can be used to avoid the need to handle IgnoredSeqNums. But then we need to consider cases where an acquisition is a no-op because an existing lock exists with the same or a higher strength.
___
*[`pkg/storage/lock_table_key_scanner.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjVry04Moi2urqPiPO:-NsjVry04Moi2urqPiPP:b-w82y9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/lock_table_key_scanner.go#L131)):*
> ```Go
> 	ctx context.Context,
> 	reader Reader,
> 	txn *enginepb.TxnMeta,
> ```

If we're moving in this direction, consider just passing in a `txnID`. We'd need to adjust the comment above.
___
*[`pkg/storage/mvcc.go` line 5902 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjWZIw6OzjJzvLXsy7:-NsjWZIw6OzjJzvLXsy8:b82or5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5902)):*
> ```Go
> }
> 
> // VerifyLock returns true if the supplied transaction holds a replicated lock
> ```

nit: there are parallels between `MVCCCheckForAcquireLock` and `MVCCAcquireLock` that are lost by putting this chunk of code between them. Let's move it below `validateLockAcquisitionStrength`.
___
*[`pkg/storage/mvcc.go` line 5913 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjXQNFCliMuLGNSHR5:-NsjXQNFCliMuLGNSHR6:b-yg6ujr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5913)):*
> ```Go
> 	}
> 	ltScanner, err := newLockTableKeyScanner(
> 		ctx, reader, txn, str, 0, 0, BatchEvalReadCategory)
> ```

We're passing in `str` and then catching `LockConflictErrors` as an indication that a matching lock doesn't exist for `txn`. I guess that works, but it's kind of subtle and it's not clear that it's needed. Should we update `lockTableKeyScanner` to work with `str = lock.None`? I _think_ the underlying `LockTableIterator` is built such that `MatchMinStr = lock.None` will not match with any other txn IDs. So the change might be as simple as:
```go
diff --git a/pkg/storage/lock_table_key_scanner.go b/pkg/storage/lock_table_key_scanner.go
index e7516f86310..1ad424e1607 100644
--- a/pkg/storage/lock_table_key_scanner.go
+++ b/pkg/storage/lock_table_key_scanner.go
@@ -62,6 +62,9 @@ func strongerOrEqualStrengths(str lock.Strength) []lock.Strength {
 // the provided lock strength.
 func minConflictLockStrength(str lock.Strength) (lock.Strength, error) {
        switch str {
+       case lock.None:
+               // Don't conflict with any locks held by other transactions.
+               return lock.None, nil
        case lock.Shared:
                return lock.Exclusive, nil
        case lock.Exclusive, lock.Intent:
```
Then we can remove the `LockConflictError` handling below.
___
*[`pkg/storage/mvcc.go` line 5938 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjYpoPE4Jav8X1USjj:-NsjYpoPE4Jav8X1USjk:bx9ylf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5938)):*
> ```Go
> 		}
> 
> 		// TODO(arul): add a comment explaining this.
> ```

Should we also check `IgnoredSeqNums`? I guess we'll need to handle cases where the lock is held but should have been rolled back and replaced by another lock. I haven't thought all the way through what that case means though.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented on pkg/storage/mvcc.go:
```diff
@@ -5891,14 +5891,60 @@ func MVCCCheckForAcquireLock(
+		}
+
+		// TODO(arul): add a comment explaining this.
+		if foundLock.Txn.Sequence <= txn.Sequence {
```
Discussion in person:
- we need a `if enginepb.TxnSeqIsIgnored(foundLock.Txn.Sequence, txn.IgnoredSeqNums) { continue } else { return true }` 
- we don't need a `foundLock.Txn.Sequence <=> txn.Sequence` comparison

nvanbenschoten commented on pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go:
```diff
@@ -541,6 +564,7 @@ func (tp *txnPipeliner) chainToInFlightWrites(ba *kvpb.BatchRequest) *kvpb.Batch
 						},
 						Txn:            meta,
 						ErrorIfMissing: true,
+						LockStrength:   w.LockStrength,
```
Discussed in person — we don't want to send QueryIntentRequests for intents+locks at ignored sequence numbers. Just delete them from `ifWrites` immediately and don't chain on to them.

arulajmani committed:
kv: add ability to verify pipelined replicated shared/exclusive locks

Previously, QueryIntent requests were only used to verify whether an
intent was successfully evaluated and replicated. This patch extends
QueryIntent request to also be able to verify whether a pipelined
shared or exclusive lock was successfully replicated or not.

Informs #117978

Release note: None

arulajmani reviewed (commented):
Marking this as ready for review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-NtxqZCm9zfig-tF5DAQ:bs5o19i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvpb/api.proto` line 1440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj__Rb1tXTjGJtWJ-B:-NtxgfSiCMyLgmAhWsPu:bt7s8m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvpb/api.proto#L1440)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to set this field in `resolveIndeterminateCommitForTxnProbe`?
</blockquote></details>

Done. We'll need to make more changes to `resolveIndeterminateCommitForTxnProbe` in the client side tracking PR as well; for now, I'm just hard coding the lock strength to `lock.Intent`.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj_wC7DNsHKwnHnuiX:-Ntxhva04MqQQNGYkzg_:b-97or9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we fix up the `NewIntentMissingError` interface so that we can pass in the missing lock?
</blockquote></details>

`NewIntentMissingError` accepts the wrong intent that was found instead of what the `QueryIntent` request was expecting. Is it meaningful in the case of replicated locks? For example, if we're querying for an exclusive lock, and an exclusive lock exists from a different transaction, does it matter? 

We'd have to jump through a few loops to return any/all locks held by other transactions if the `QueryIntent` didn't find the expected replicated lock. I'm not sure it's worth it. Wdyt?
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxpmPz7F4wbsUTx6vJ:-NtxpmPz7F4wbsUTx6vK:bikzik2) ([raw file](https://github.com/cockroachdb/cockroach/blob/de9994b3dc61b0f2bb9310d599f31c6a75ff03c3/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L50)):*
> ```Go
> 	// TODO(arul): add a test.
> 	//
> 	// TODO(XXX): Do we really need this? We're saying that we don't need to
> ```

@nvanbenschoten wanted to call this bit out for discussion. I may be missing something here.
___
*[`pkg/storage/lock_table_key_scanner.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjVry04Moi2urqPiPO:-NtwS7fUDc_yG6H0f3-7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/lock_table_key_scanner.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we're moving in this direction, consider just passing in a `txnID`. We'd need to adjust the comment above.
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 5913 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjXQNFCliMuLGNSHR5:-NtxffSN3TAV-YShxO1y:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5913)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're passing in `str` and then catching `LockConflictErrors` as an indication that a matching lock doesn't exist for `txn`. I guess that works, but it's kind of subtle and it's not clear that it's needed. Should we update `lockTableKeyScanner` to work with `str = lock.None`? I _think_ the underlying `LockTableIterator` is built such that `MatchMinStr = lock.None` will not match with any other txn IDs. So the change might be as simple as:
```go
diff --git a/pkg/storage/lock_table_key_scanner.go b/pkg/storage/lock_table_key_scanner.go
index e7516f86310..1ad424e1607 100644
--- a/pkg/storage/lock_table_key_scanner.go
+++ b/pkg/storage/lock_table_key_scanner.go
@@ -62,6 +62,9 @@ func strongerOrEqualStrengths(str lock.Strength) []lock.Strength {
 // the provided lock strength.
 func minConflictLockStrength(str lock.Strength) (lock.Strength, error) {
        switch str {
+       case lock.None:
+               // Don't conflict with any locks held by other transactions.
+               return lock.None, nil
        case lock.Shared:
                return lock.Exclusive, nil
        case lock.Exclusive, lock.Intent:
```
Then we can remove the `LockConflictError` handling below.
</blockquote></details>

Done. 
___
*[`pkg/storage/mvcc.go` line 5938 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjYpoPE4Jav8X1USjj:-NtwRWCvDE3SlZQdzROq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5938)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we also check `IgnoredSeqNums`? I guess we'll need to handle cases where the lock is held but should have been rolled back and replaced by another lock. I haven't thought all the way through what that case means though.
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 5939 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#gh-1525458076:-NtwRGdNESgTM8U1ITJ-:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/storage/mvcc.go#L5939)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Discussion in person:
- we need a `if enginepb.TxnSeqIsIgnored(foundLock.Txn.Sequence, txn.IgnoredSeqNums) { continue } else { return true }` 
- we don't need a `foundLock.Txn.Sequence <=> txn.Sequence` comparison
</blockquote></details>

Done. 
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go` line 567 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#gh-1525472531:-NtxgvTVAo4bfqXYAnA6:bl20twc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go#L567)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Discussed in person — we don't want to send QueryIntentRequests for intents+locks at ignored sequence numbers. Just delete them from `ifWrites` immediately and don't chain on to them.
</blockquote></details>

More of a note to you, now that you're doing the client side piece.
___
*[`pkg/roachpb/data.go` line 1004 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjWJ2Y9RMGbLPL0-JQ:-NtxgJ0YDxmmmUaZHBEv:b6z9o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/roachpb/data.go#L1004)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This doesn't feel like it belongs as a method on `Transaction`. Rather, it feels like it should be a method on `MVCCWriteOptions`.
</blockquote></details>

Got rid of this.
___
*[`pkg/roachpb/data.proto` line 630 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NsjVV6HBJSJOSXRmtxi:-Ntxq15M4sMxmvCg1KwI:bvwma4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/roachpb/data.proto#L630)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This will only be set if `lock_strength` is Intent, right?

EDIT: or maybe not. Maybe this can be used to avoid the need to handle IgnoredSeqNums. But then we need to consider cases where an acquisition is a no-op because an existing lock exists with the same or a higher strength.
</blockquote></details>

No longer part of this PR, but you're right we're not using this on the server. It's up to us if we want to special case this tracking on the client or if we want to track this for all lock strengths but just never use it during evaluation.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
Nice job on this! Leaving a handful of comments, but none are substantial or should require much effort to address. Will stamp right after they're addressed.

> Reviewed 8 of 15 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-NtxzQHV-aXfTFZvMVhg:b-rkqt41)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)

___
*[`pkg/kv/kvpb/api.proto` line 1444 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxt817A26MPQ5x3nB1:-Ntxt817A26MPQ5x3nB2:bg4tzpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvpb/api.proto#L1444)):*
> ```proto
>   // with the supplied lock strength or something stronger at the sequence
>   // number.
>   kv.kvserver.concurrency.lock.Strength lock_strength = 4;
> ```

nit: we're going to rename this request type to `QueryLock` at some point, so we might as well just name this `strength` to avoid `QueryLock.LockStrength`.
___
*[`pkg/kv/kvpb/api.proto` line 1450 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxtXrr11II5G_DFsjM:-NtxtXrr11II5G_DFsjN:b2wxnnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvpb/api.proto#L1450)):*
> ```proto
>   // ignored will be treated as "not found"; that's because they can be removed
>   // at any time.
>   repeated storage.enginepb.IgnoredSeqNumRange ignored_seqnums = 5
> ```

Now that I see this spelled out, it probably does make sense to add validation to `cmd_query_intent.go` that `!enginepb.TxnSeqIsIgnored(args.Txn.Sequence, args.IgnoredSeqNums)`. Otherwise, return an `AssertionFailed` error.
___
*[`pkg/kv/kvserver/replica_tscache.go` line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxuSQJ3QaNQwawjzwH:-NtxuSQJ3QaNQwawjzwI:b-98h2j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/replica_tscache.go#L315)):*
> ```Go
> 				addToTSCache(start, end, t.Txn.WriteTimestamp, uuid.UUID{})
> 			}
> 			// NB: If this QueryIntentRequest was querying a replicated lock instead
> ```

Given all this, should we make the previous logic conditional on `t.Strength == lock.Intent` (with the corresponding mixed version logic, which we might want to pull behind a `StrengthOrDefault` method on `QueryIntentRequest`)?
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj_wC7DNsHKwnHnuiX:-Ntxvmxg6MFpSrLfCQ1I:b-4k9be1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L149)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

`NewIntentMissingError` accepts the wrong intent that was found instead of what the `QueryIntent` request was expecting. Is it meaningful in the case of replicated locks? For example, if we're querying for an exclusive lock, and an exclusive lock exists from a different transaction, does it matter? 

We'd have to jump through a few loops to return any/all locks held by other transactions if the `QueryIntent` didn't find the expected replicated lock. I'm not sure it's worth it. Wdyt?
</blockquote></details>

Oh you're right, I was confused and thinking about `RefreshFailedError` instead. Agreed that we want to pass `nil`.

Should we hoist the `intent` variable up a scope and then avoid the duplicate `!reply.FoundIntent && args.ErrorIfMissing` checks?
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxpmPz7F4wbsUTx6vJ:-Ntxv-u1BWy91k8x45-G:bo9gjsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/de9994b3dc61b0f2bb9310d599f31c6a75ff03c3/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L50)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

@nvanbenschoten wanted to call this bit out for discussion. I may be missing something here.
</blockquote></details>

This isn't to prevent the replicated locks from landing after the query intent request, it's to ensure that the query intent request waits on the lock acquisition's replication. Otherwise, it would slip ahead and fail to find the replicated lock.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxupkj9oNsW0vXnVT7:-Ntxupkj9oNsW0vXnVT8:b-p20m2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L57)):*
> ```Go
> 	// them. Does that mean this latching isn't required?
> 	txnID := req.(*kvpb.QueryIntentRequest).Txn.ID
> 	latchSpans.AddNonMVCC(spanset.SpanReadWrite, roachpb.Span{
> ```

`s/SpanReadWrite/SpanReadOnly/`
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxvM9u58uAfmz7_zsU:-NtxvM9u58uAfmz7_zsV:banrc9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L100)):*
> ```Go
> 	// TODO(arul): We should be able to remove the lock.None case once
> 	// compatibility with 24.1 is no longer an issue.
> 	if args.LockStrength == lock.Intent || args.LockStrength == lock.None {
> ```

See comment above about hiding this behind a method to centralize the compatibility logic.
___
*[`pkg/kv/kvserver/txnrecovery/manager.go` line 210 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxwHf_BQpTSWqiV81k:-NtxwHf_BQpTSWqiV81l:b-v0yczw) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/txnrecovery/manager.go#L210)):*
> ```Go
> 			},
> 			Txn:            meta,
> 			LockStrength:   lock.Intent,
> ```

Want to leave me a TODO here? Just `TODO(nvanbenschoten): pass in the correct strength.`
___
*[`pkg/storage/lock_table_key_scanner.go` line 145 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxwa_l4-bOotaRtqDm:-Ntxwa_l4-bOotaRtqDn:bsvusny) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/lock_table_key_scanner.go#L145)):*
> ```Go
> 	readCategory ReadCategory,
> ) (*lockTableKeyScanner, error) {
> 	if txnID.Equal(uuid.UUID{}) && str == lock.None {
> ```

We can probably skip this. It's already checked in `LockTableIteratorOptions.validate`, so no need for the noise here.
___
*[`pkg/storage/mvcc.go` line 5967 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxxzgV5Ql1N0PIKOJ9:-NtxxzgV5Ql1N0PIKOJA:b-njfezu) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/mvcc.go#L5967)):*
> ```Go
> 		return err
> 	}
> 	txnID := txn.ID
> ```

small nit: might as well just inline this instead of pulling out the variable, like you do below. It's saving one character.
___
*[`pkg/storage/mvcc.go` line 6165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxyFdD905Y88gKVydL:-NtxyFdD905Y88gKVydM:b-seidgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/mvcc.go#L6165)):*
> ```Go
> 		// nothing stopping another request from rolling back the lock even though
> 		// it exists right now.
> 		if !enginepb.TxnSeqIsIgnored(foundLock.Txn.Sequence, ignoredSeqNums) {
> ```

In my opinion, structuring this like `if enginepb.TxnSeqIsIgnored(... { continue }` and then `return true` below would be easier to read. That way the happy case is split out from the various exceptional cases, which each result in an early `continue`.
___
*[`pkg/storage/testdata/mvcc_histories/verify_locks` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxytq_DOKPVkqFbiJP:-Ntxytq_DOKPVkqFbiJQ:b-j0rq55) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/testdata/mvcc_histories/verify_locks#L59)):*
> ```
> with t=A
>   txn_step seq=10
>   verify_lock k=k1 str=shared
> ```

Should we test the base case somewhere of a key with no lock?
___
*[`pkg/storage/testdata/mvcc_histories/verify_locks` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxyzbQ98dGgsJ06swx:-NtxyzbQ98dGgsJ06swy:bdxho23) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/testdata/mvcc_histories/verify_locks#L59)):*
> ```
> with t=A
>   txn_step seq=10
>   verify_lock k=k1 str=shared
> ```

Also, should we test the case where the key has an intent on it?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-Nu-VvlW-_emEsZLyM93:bb72pd0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_tscache.go` line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxuSQJ3QaNQwawjzwH:-Nu-Mc_sCCIYrQhca24f:beic5y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/replica_tscache.go#L315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Given all this, should we make the previous logic conditional on `t.Strength == lock.Intent` (with the corresponding mixed version logic, which we might want to pull behind a `StrengthOrDefault` method on `QueryIntentRequest`)?
</blockquote></details>

Done. Reworded the comment here too to work better with the new structure.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nsj_wC7DNsHKwnHnuiX:-Nu-N8dlDBlHM1R2p8lB:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5269438f626350e8ab325506ad1a4169636af16b/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L149)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Oh you're right, I was confused and thinking about `RefreshFailedError` instead. Agreed that we want to pass `nil`.

Should we hoist the `intent` variable up a scope and then avoid the duplicate `!reply.FoundIntent && args.ErrorIfMissing` checks?
</blockquote></details>

Good call, done.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxpmPz7F4wbsUTx6vJ:-Nu-I76jCHEkk0Ya1l1I:b2qd9nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/de9994b3dc61b0f2bb9310d599f31c6a75ff03c3/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L50)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This isn't to prevent the replicated locks from landing after the query intent request, it's to ensure that the query intent request waits on the lock acquisition's replication. Otherwise, it would slip ahead and fail to find the replicated lock.
</blockquote></details>

Ah right!
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxupkj9oNsW0vXnVT7:-Nu-IB7fBGH8j_5-DxHZ:b-snvfqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L57)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/SpanReadWrite/SpanReadOnly/`
</blockquote></details>

Copy pasta. Fixed.
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxvM9u58uAfmz7_zsU:-Nu-MgiV2aTBRZzwg3d_:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L100)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

See comment above about hiding this behind a method to centralize the compatibility logic.
</blockquote></details>

Done. 
___
*[`pkg/storage/mvcc.go` line 5967 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxxzgV5Ql1N0PIKOJ9:-Nu-HPa7ENClZtQ8azCT:byapuf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/mvcc.go#L5967)):*
> It's saving one character.

Every little bit counts, but okay :joy: 
___
*[`pkg/storage/mvcc.go` line 6165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxyFdD905Y88gKVydL:-Nu-HzhT0JuB4i6biJnR:b-kdffca) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/mvcc.go#L6165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In my opinion, structuring this like `if enginepb.TxnSeqIsIgnored(... { continue }` and then `return true` below would be easier to read. That way the happy case is split out from the various exceptional cases, which each result in an early `continue`.
</blockquote></details>

Agreed. Also not sure what I was thinking with the "continue" there :P 
___
*[`pkg/storage/testdata/mvcc_histories/verify_locks` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxyzbQ98dGgsJ06swx:-Nu-VtZv3XgUjgjxjEz8:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/testdata/mvcc_histories/verify_locks#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Also, should we test the case where the key has an intent on it?
</blockquote></details>

Done
___
*[`pkg/storage/testdata/mvcc_histories/verify_locks` line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Ntxytq_DOKPVkqFbiJP:-Nu-VuWhA0Wa-OdNNQF_:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/storage/testdata/mvcc_histories/verify_locks#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we test the base case somewhere of a key with no lock?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-Nu-XYLT0Ndt2V5W10Vh:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)

___
*[`pkg/kv/kvpb/api.proto` line 1450 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxtXrr11II5G_DFsjM:-Nu-XV1pAeKpPuTVmhYt:b6f5pz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvpb/api.proto#L1450)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Now that I see this spelled out, it probably does make sense to add validation to `cmd_query_intent.go` that `!enginepb.TxnSeqIsIgnored(args.Txn.Sequence, args.IgnoredSeqNums)`. Otherwise, return an `AssertionFailed` error.
</blockquote></details>

Did we want to add something like this to `cmd_query_intent.go`?
___
*[`pkg/kv/kvserver/replica_tscache.go` line 313 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nu-WsUX5LziGgfKan6r:-Nu-WsUX5LziGgfKan6s:b-o7y2yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/27138c37bd9bf9dba168dc2b4802850488e784dd/pkg/kv/kvserver/replica_tscache.go#L313)):*
> ```Go
> 			// the request evaluated successfully (so it can't land later) -- it's
> 			// only checking whether the replication succeeded or not.
> 			if req.(*kvpb.QueryIntentRequest).StrengthOrDefault() == lock.Intent {
> ```

`s/req.(*kvpb.QueryIntentRequest)/t/`
___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 46 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nu-X2Ki-Vk0cGVPX21H:-Nu-X2Ki-Vk0cGVPX21I:bvry3hl) ([raw file](https://github.com/cockroachdb/cockroach/blob/27138c37bd9bf9dba168dc2b4802850488e784dd/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L46)):*
> ```Go
> 	latchSpans.AddNonMVCC(spanset.SpanReadOnly, req.Header().Span())
> 	// They also acquire a read latch on the per-transaction local key that all
> 	// replicated shared lock acquisitions acquire latches on to isolate them
> ```

"to isolate them To isolate themselves"


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Added the version gate as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-Nu-m0px9Ie7nndrLtj4:b-h796ph)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @jbowens, and @nvanbenschoten)

___
*[`pkg/kv/kvpb/api.proto` line 1450 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-NtxtXrr11II5G_DFsjM:-Nu-kXQ88MMq8AF36ljA:bg24ttv) ([raw file](https://github.com/cockroachdb/cockroach/blob/70b7031569d9ef1eb74817c7bd56ccdcff2227eb/pkg/kv/kvpb/api.proto#L1450)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did we want to add something like this to `cmd_query_intent.go`?
</blockquote></details>

Sorry for missing the comment earlier. Done.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> once CI is green

> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-Nu-nWPD6hwsiV9RoVc2:bxdwmhn)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @jbowens)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nu-nHW25BwR8h4qkM1M:-Nu-nHW25BwR8h4qkM1N:b7mhknp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c29ac1840027f2406c24841ac94a965c4d3c42fb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L91)):*
> ```Go
> 	}
> 
> 	if !enginepb.TxnSeqIsIgnored(args.Txn.Sequence, args.IgnoredSeqNums) {
> ```

Shouldn't this assertion fire `if enginepb.TxnSeqIsIgnored(args.Txn.Sequence, args.IgnoredSeqNums)` (no negation)?


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Looks like CI is hitting:
```
pkg/kv/kvserver/batcheval/cmd_query_intent.go:92:27: github.com/cockroachdb/errors.AssertionFailedf format %s has arg args.Txn.Sequence of wrong type github.com/cockroachdb/cockroach/pkg/storage/enginepb.TxnSeq (printf)
```

arulajmani reviewed (commented):
Thanks for the quick reviews and all the help here @nvanbenschoten! 

bors r=nvanbenschoten

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-:-Nu05uAc9vbkkv4UOo3R:bqkjek6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens and @nvanbenschoten)

___
*[`pkg/kv/kvserver/batcheval/cmd_query_intent.go` line 91 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/119933#-Nu-nHW25BwR8h4qkM1M:-Nu05r9I2YFTFPm_c2Xn:b-z3co3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c29ac1840027f2406c24841ac94a965c4d3c42fb/pkg/kv/kvserver/batcheval/cmd_query_intent.go#L91)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Shouldn't this assertion fire `if enginepb.TxnSeqIsIgnored(args.Txn.Sequence, args.IgnoredSeqNums)` (no negation)?
</blockquote></details>

It should have. Trying to work too fast :/ 


<!-- Sent from Reviewable.io -->
